runtime/core/testing_util/error_matchers.h

```diff
@@ -0,0 +1,343 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+/**
+ * @file
+ * Testing utilities for working with `executorch::runtime::Result<T>` and
+ * `executorch::runtime::Error`. Provides matchers similar to `absl::StatusOr`
+ * and `absl::Status`.
+ *
+ * Defines the following utilities:
+ *
+ *   ===============
+ *   `IsOkAndHolds(m)`
+ *   ===============
+ *
+ *   This gMock matcher matches a Result<T> value whose error is Ok
+ *   and whose inner value matches matcher m.  Example:
+ *
+ *   ```
+ *   using ::testing::MatchesRegex;
+ *   using ::executorch::runtime::testing::IsOkAndHolds;
+ *   ...
+ *   executorch::runtime::Result<string> maybe_name = ...;
+ *   EXPECT_THAT(maybe_name, IsOkAndHolds(MatchesRegex("John .*")));
+ *   ```
+ *
+ *   ===============
+ *   `ErrorIs(Error::error_code)`
+ *   ===============
+ *
+ *   This gMock matcher matches a Result<T> value whose error matches
+ *   the given error matcher. Example:
+ *
+ *   ```
+ *   using ::executorch::runtime::testing::ErrorIs;
+ *   ...
+ *   executorch::runtime::Result<string> maybe_name = ...;
+ *   EXPECT_THAT(maybe_name, ErrorIs(Error::InvalidArgument));
+ *   ```
+ *
+ *   ===============
+ *   `IsOk()`
+ *   ===============
+ *
+ *   Matches an `executorch::runtime::Result<T>` value whose error value
+ *   is `executorch::runtime::Error::Ok`.
+ *
+ *   Example:
+ *   ```
+ *   using ::executorch::runtime::testing::IsOk;
+ *   ...
+ *   executorch::runtime::Result<string> maybe_name = ...;
+ *   EXPECT_THAT(maybe_name, IsOk());
+ *   ```
+ */
+
+#pragma once
+
+#include <ostream>
+#include <utility>
+
+#include <gmock/gmock-matchers.h>
+#include <gtest/gtest-matchers.h>
+
+#include <executorch/runtime/core/error.h>
+#include <executorch/runtime/core/result.h>
+
+/**
+ * Unwrap a Result to obtain its value. If the Result contains an error,
+ * fail the test with ASSERT_TRUE.
+ *
+ * This macro is useful for test code where you want to extract the value
+ * from a Result and fail the test if the Result contains an error.
+ *
+ * Example usage:
+ * ```
+ *   Result<int> maybe_value = GetSomeValue();
+ *   int value = ASSERT_OK_AND_UNWRAP(maybe_value);
+ *   // Use value...
+ * ```
+ *
+ * @param[in] result__ Expression yielding the Result to unwrap.
+ */
+#define ASSERT_OK_AND_UNWRAP(result__) \
+  ({                                   \
+    auto&& et_result__ = (result__);   \
+    ASSERT_TRUE(et_result__.ok());     \
+    std::move(*et_result__);           \
+  })
+
+namespace executorch {
+namespace runtime {
+namespace testing {
+namespace internal {
+namespace {
+
+constexpr const char* toString(const Error error) {
```

Comment:
Can you put this in Error? Would be easier for future changes to Error also change this API.